src/test/java/com/google/firebase/remoteconfig/TemplateTest.java

```diff
@@ -153,4 +166,192 @@ public void testEquality() {
+    t.getParameters()
+            .put("with_inApp", new Parameter().setDefaultValue(ParameterValue.inAppDefault()));
+    String jsonString = t.toJSON();
+
+    assertEquals("{\"conditions\":[],\"parameterGroups\":{},"
+            + "\"parameters\":{\"with_value\":{\"conditionalValues\":{},"
+            + "\"defaultValue\":{\"value\":\"hello\"}},\"with_inApp\":{\"conditionalValues\":{},"
+            + "\"defaultValue\":{\"useInAppDefault\":true}}}}", jsonString);
+  }
+
+  @Test
+  public void testToJSONWithEtag() {
+    String jsonString = new Template("etag-12345").toJSON();
+
+    assertEquals("{\"conditions\":[],\"etag\":\"etag-12345\",\"parameterGroups\":{},"
+            + "\"parameters\":{}}", jsonString);
+  }
+
+  @Test
+  public void testToJSONWithEtagAndConditions() {
+    String jsonString = new Template("etag-0010201")
+            .setConditions(CONDITIONS).toJSON();
+
+    assertEquals("{\"conditions\":[{\"expression\":\"exp ios\",\"name\":\"ios_en\","
+            + "\"tagColor\":\"INDIGO\"},{\"expression\":\"exp android\","
+            + "\"name\":\"android_en\"}],\"etag\":\"etag-0010201\",\"parameterGroups\":{},"
+            + "\"parameters\":{}}", jsonString);
+  }
+
+  @Test
+  public void testToJSONWithVersion() {
+    Version version = new Version(new TemplateResponse.VersionResponse()
+            .setDescription("template version")
+            .setUpdateTime("2020-12-08T15:49:51.887878Z")
+            .setUpdateUser(new TemplateResponse.UserResponse().setEmail("user@user.com"))
+            .setLegacy(false)
+            .setUpdateType("INCREMENTAL_UPDATE")
+            .setRollbackSource("26")
+            .setVersionNumber("34")
+            .setUpdateOrigin("ADMIN_SDK_NODE")
+    );
+    String jsonString = new Template().setVersion(version).toJSON();
+
+    assertEquals("{\"conditions\":[],\"parameterGroups\":{},\"parameters\":{},"
+            + "\"version\":{\"description\":\"template version\",\"legacy\":false,"
+            + "\"rollbackSource\":\"26\",\"updateOrigin\":\"ADMIN_SDK_NODE\","
+            + "\"updateTime\":\"Tue, 08 Dec 2020 15:49:51 UTC\","
+            + "\"updateType\":\"INCREMENTAL_UPDATE\",\"updateUser\":{"
+            + "\"email\":\"user@user.com\"},\"versionNumber\":\"34\"}}", jsonString);
+  }
+
+  @Test
+  public void testToJSONAndFromJSON() throws FirebaseRemoteConfigException {
+    Template originalTemplate = new Template();
+    Template otherTemplate = Template.fromJSON(originalTemplate.toJSON());
+
+    assertEquals(originalTemplate, otherTemplate);
+
+    Version expectedVersion = Version.withDescription("promo version");
+    originalTemplate = new Template("etag-0010201")
+            .setParameters(PARAMETERS)
+            .setConditions(CONDITIONS)
+            .setParameterGroups(PARAMETER_GROUPS)
+            .setVersion(expectedVersion);
+    Template template = Template.fromJSON(originalTemplate.toJSON());
+
+    assertEquals("etag-0010201", template.getETag());
```

Comment:
Why not run assertEquals on the template instances?

Reply:
Condition arrays are not quite the same for a direct equality check. The original template contains a condition that does not have a tag color. After the serialization -> deserialization process that tag color will be set to `TagColor.UNSPECIFIED`.

Reply:
Should we fix that so the behavior remains consistent?

Reply:
I think we should. Updated the code not to serialize the tag color as `TagColor.UNSPECIFIED` when `null` or empty.